mlflow/shap.py

```diff
@@ -1,19 +1,97 @@
 from contextlib import contextmanager
 import os
 import tempfile
+import yaml
 
 import numpy as np
 
 import mlflow
+import types
+from mlflow import pyfunc
+from mlflow.exceptions import MlflowException
 from mlflow.utils.annotations import experimental
 from mlflow.utils.uri import append_to_uri_path
+from mlflow.models import Model
 
+from mlflow.tracking.artifact_utils import _download_artifact_from_uri
+from mlflow.models.model import MLMODEL_FILE_NAME
+from mlflow.models.signature import ModelSignature
+from mlflow.models.utils import ModelInputExample, _save_example
+from mlflow.utils.environment import _mlflow_conda_env
+from mlflow.utils.model_utils import _get_flavor_configuration
+from mlflow.protos.databricks_pb2 import RESOURCE_ALREADY_EXISTS
+from mlflow.tracking._model_registry import DEFAULT_AWAIT_MAX_SLEEP_SECONDS
+
+FLAVOR_NAME = "shap"
 
 _MAXIMUM_BACKGROUND_DATA_SIZE = 100
 _DEFAULT_ARTIFACT_PATH = "model_explanations_shap"
 _SUMMARY_BAR_PLOT_FILE_NAME = "summary_bar_plot.png"
 _BASE_VALUES_FILE_NAME = "base_values.npy"
 _SHAP_VALUES_FILE_NAME = "shap_values.npy"
+_UNKNOWN_MODEL_FLAVOR = "unknown"
+_UNDERLYING_MODEL_SUBPATH = "underlying_model"
+
+
+def get_underlying_model_flavor(model):
+    """
+    Find the underlying models flavor.
+
+    :param model: underlying model of the explainer.
+    """
+
+    # checking if underlying model is wrapped
+
+    if hasattr(model, "model"):
+        unwrapped_model = model.model
+        if isinstance(unwrapped_model, types.FunctionType):
+            return "python_function"
```

Comment:
Should we avoid using `python_function` since it sounds like MLflow's `PyFuncModel`?

https://github.com/mlflow/mlflow/blob/ebd4e5ed2e12055d5b1d3ff3c740fd838301f616/mlflow/pyfunc/__init__.py#L232

Reply:
Okay so this was meant to check if the underlying model was PyFunc Flavor (python function), but we're currently not handling that flavor in the save_model function, so I removed this check. For now its treated as UNKNOWN_MODEL_FLAVOR, and SHAP tries to serialize it itself.